frontend/src/components/ChatRoom/UserList.tsx

```diff
@@ -1,64 +1,89 @@
-import { useEffect, useState } from 'react';
+import ChannelInvitationButton from './ChannelInvitationButton';
+import ChannelUpdateButton from './ChannelUpdateButton';
 import instance from '../../util/Axios';
+import { useEffect, useState } from 'react';
 import User from '../User/User';
-import { ErrorAlert } from '../../util/Alert';
-import ChannelInvitationButton from './ChannelInvitationButton';
 import { useRecoilValue } from 'recoil';
-import { myIdState } from '../../util/Recoils';
 import { useSocketOn } from '../hooks/SocketOnHooks';
-import ChannelUpdateButton from './ChannelUpdateButton';
+import { ErrorAlert } from '../../util/Alert';
+import { myIdState } from '../../util/Recoils';
+import { socket } from '../../util/Socket';
+import { useNavigate } from 'react-router-dom';
+import UserLeftButton from './UserLeftButton';
 
 interface Props {
   id: string;
 }
 
-// isReadonlyDm : ??
-// true일 때 읽기만 가능한 DM
-// false : DM인데 가능
-// null : 단톡
-// socket on roleChanged
-// memberJoin
-// memberLeft
-// 이벤트 발생 시 channelMembers 업데이트
-
 function UserList(props: Props) {
-  const [channelInfo, setChannelInfo] = useState<{
-    channelMembers: { id: number; role: 'owner' | 'admin' | 'member' }[];
-    isReadonlyDm: boolean | null;
-  }>({
-    channelMembers: [],
-    isReadonlyDm: null,
-  });
+  const [channelMembers, setChannelMembers] = useState<
+    { id: number; role: 'owner' | 'admin' | 'member' }[]
+  >([]);
+  const [isReadonlyDm, setIsReadonlyDm] = useState<boolean | null>(null);
   const [myRole, setMyRole] = useState<'owner' | 'admin' | 'member'>('member');
   const myId = useRecoilValue(myIdState);
+  const navigate = useNavigate();
 
   useSocketOn();
 
   useEffect(() => {
     instance
       .get(`/chats/${props.id}`)
       .then(result => {
-        setChannelInfo(result.data);
+        setChannelMembers(result.data.channelMembers);
+        setIsReadonlyDm(result.data.isReadonlyDm);
         setMyRole(
           result.data.channelMembers.find(
             ({ id }: { id: number }) => id === myId,
           )?.role || 'member',
         );
+        socket.on('memberJoin', data => {
+          setChannelMembers(prev => [
+            ...prev,
+            { id: data.joinedMember, role: 'member' },
+          ]);
+        });
+        socket.on('memberLeft', data => {
+          if (data.isOwner) {
+            ErrorAlert(
+              '방장이 채팅방을 나갔습니다.',
+              '채팅방이 삭제되었습니다',
+            );
+            navigate('/chats');
+          }
+          setChannelMembers(
+            channelMembers.filter(member => member.id !== data.leftMember),
+          );
+        });
+        socket.on('roleChanged', data => {
+          let copy = [...channelMembers];
+          let index = copy.findIndex(
```

Comment:
둘 다 const 여도 될 것 같습니다.
```suggestion
          const copy = [...channelMembers];
          const index = copy.findIndex(
```